### ISSUE TITLE
Constants and collections clash without error or documentation.

### ISSUE THREAD

**Issue by CohenCyril** [coq/coq#13296]
## Description of the problem

The following code does not raise an error, and the behavior is undocumented. 

```coq
Section Test.
Variables (b : bool) (n : nat) (All : unit).
Collection n := b.

Lemma foo : True.
Proof using n.
trivial.
Qed.

Lemma bar : True.
Proof using All.
trivial.
Qed.
End Test.

Check foo : bool -> True.
Check bar : bool -> nat -> unit -> True.
```

I detected this problem while looking at the new attribute syntax `using` in #13183, and suggesting that we switch to the syntax `{x y z}` for single names and `c` for collections, as in commands `Proof using ({x y z} + c) - {y}` and `Collection c' := ({x y z} + c) - {y}`, and attribute `#[using="({x y z} + c) - {y}"]`.

I suggest, if it is doable, we keep backward compatibility the current syntax with a warning ("deprecated syntax collection") when there is no ambiguity, and raise an error otherwise.

#### Coq Version
Coq 8.10 at least but most probably from Coq 8.5 and above.

**Comment by CohenCyril** [comment]
CC @Zimmi48  @gares

**Comment by gares** [comment]
Can't we just give an error if there is a clash? No syntax change, we could warn in one release and error in the next.
I mean, I see no use cases for using the same name for both a variable and a collection thereof.

**Comment by CohenCyril** [comment]
I agree, but it does mean that `All` is now forbidden as a variable name...

**Comment by CohenCyril** [comment]
> I mean, I see no use cases for using the same name for both a variable and a collection thereof.

**Comment by gares** [comment]
As well as `Type`

**Comment by herbelin** [comment]
By the way, if I may give an innocent outsider comment, when I saw `a*`, I first thought that it meant all variables whose name starts with `a`. Then I saw `Type*` and I started to be confused. (I confess that I read the documentation only afterwards.)

**Comment by Zimmi48** [comment]
I would also favor changing the behavior instead of the syntax.

Are hard errors necessary? Can't this simply be warnings? Every time one would invoke a name which is both a collection name (including the built-in `All`) and a variable name in a `using` or a `Collection` context, a warning would be raised noting the ambiguity, specifying what is chosen (the collection) and that it is discouraged to give the same names to section variables and collections.

The only thing I would make into a hard error would be `Collection All :=` because it has strictly no use case (`All` will always be interpreted as the built-in collection).

**Comment by gares** [comment]
Yup, the code interpreting `Collection <name> := ...` should check if a section variable `<name>` exists and give a warning. It should also raise an error if `<name>` is `All`.

**Comment by CohenCyril** [comment]
I still do not understand where the error or warning about `Variable All` should occur. Should `All` be forbidden as a variable name?

**Comment by Zimmi48** [comment]
> Yup, the code interpreting `Collection <name> := ...` should check if a section variable `<name>` exists and give a warning. It should also raise an error if `<name>` is `All`.

This is not where I imagined the warning (since nothing prevents introducing the clashing section variable after the collection of the same name). I was suggesting to raise the warning everywhere `foo` is interpreted as a collection name rather than an existing variable name.

> I still do not understand where the error or warning about `Variable All` should occur. Should `All` be forbidden as a variable name?

I would forbid `Collection All :=`, but I wouldn't forbid `Variable All`.

**Comment by gares** [comment]
> I was suggesting to raise the warning everywhere `foo` is interpreted as a collection name rather than an existing variable name.

You can do it at `Collection <name>` time and as well at `Proof using <name>` time (IIRC the code is in the same file).
